spectrogram-feature-extraction.html

```diff
@@ -0,0 +1,468 @@
+    // the full length of the buffer.
+    pos = 0;
+    end = buffer.byteLength;
+  }
+
+  let v = new DataView(buffer);
+
+  function u8() {
+    let x = v.getUint8(pos);
+    pos++;
+    return x;
+  }
+
+  function u16() {
+    let x = v.getUint16(pos, true);
+    pos += 2;
+    return x;
+  }
+
+  function u32() {
+    let x = v.getUint32(pos, true);
+    pos += 4;
+    return x;
+  }
+
+  function string(len) {
+    let str = '';
+    for (let i = 0; i < len; ++i)
+      str += String.fromCharCode(u8());
+    return str;
+  }
+
+  if (string(4) !== 'RIFF')
+    throw new TypeError('Invalid WAV file');
+  u32();
+  if (string(4) !== 'WAVE')
+    throw new TypeError('Invalid WAV file');
+
+  let fmt;
+
+  while (pos < end) {
+    let type = string(4);
+    let size = u32();
+    let next = pos + size;
+    switch (type) {
+    case 'fmt ':
+      let formatId = u16();
+      if (formatId !== 0x0001 && formatId !== 0x0003)
+        throw new TypeError('Unsupported format in WAV file: ' + formatId.toString(16));
+      fmt = {
+        format: 'lpcm',
+        floatingPoint: formatId === 0x0003,
+        channels: u16(),
+        sampleRate: u32(),
+        byteRate: u32(),
+        blockSize: u16(),
+        bitDepth: u16(),
+      };
+      break;
+    case 'data':
+      if (!fmt)
+        throw new TypeError('Missing "fmt " chunk.');
+      let samples = Math.floor(size / fmt.blockSize);
+      let channels = fmt.channels;
+      let sampleRate = fmt.sampleRate;
+      let channelData = [];
+      for (let ch = 0; ch < channels; ++ch)
+        channelData[ch] = new Float32Array(samples);
+      pcm16(buffer, pos, channelData, channels, samples);
+      return {
+        sampleRate: sampleRate,
+        channelData: channelData
+      };
+      break;
+    }
+    pos = next;
+  }
+}
+
+%% md
+# Fecth an audio file
+
+let's define a function that fetches an audio file from an url and returns a promise with our arrayBuffer
+
+%% js
+function getAudioBuffer(url) {
+  return new Promise((resolve, reject) => {
+    const request = new XMLHttpRequest();
+    request.open('GET', url, true);
+    request.responseType = 'arraybuffer';
+    request.onreadystatechange = function(event) {
+      if (request.readyState == 4) {
+        if (request.status == 200 || request.status == 0) {
+          resolve(request.response);
+        } else {
+          reject({error: '404 Not found'});
+        }
+      }
+    };
+    request.send(null);
+  })
+}
+
+%% md
+## Fecth and extract
+
+Now we fetch our audio from some URL, decode and extract the feature using our python function
+
+%% js
+let url = 'https://raw.githubusercontent.com/gabrielfreire/neuralnet.js/more_experimentation/www/data/example.wav';
+getAudioBuffer(url).then((arrayBuffer) => {
+  let audioBuffer = decode(arrayBuffer);
+  let buff = audioBuffer.channelData[0];
+  // import python function
+  let getSpectrogram = pyodide.pyimport('spectrogram');
```

Comment:
For readability, maybe use the same variable name on both the Python and Javascript sides, since they refer to the same thing.